src/EditorFeatures/Core/Implementation/CodeDefinitionWindow/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,193 @@
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var workspace = document.Project.Solution.Workspace;
+            var navigableItems = await GoToDefinitionHelpers.GetDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
+            if (navigableItems?.Any() == true)
+            {
+                var navigationService = workspace.Services.GetRequiredService<IDocumentNavigationService>();
+
+                var builder = new ArrayBuilder<CodeDefinitionWindowLocation>();
+                foreach (var item in navigableItems)
+                {
+                    if (await navigationService.CanNavigateToSpanAsync(workspace, item.Document.Id, item.SourceSpan, cancellationToken).ConfigureAwait(false))
+                    {
+                        var text = await item.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                        var linePositionSpan = text.Lines.GetLinePositionSpan(item.SourceSpan);
+
+                        if (item.Document.FilePath != null)
+                        {
+                            builder.Add(new CodeDefinitionWindowLocation(item.DisplayTaggedParts.JoinText(), item.Document.FilePath, linePositionSpan.Start));
+                        }
+                    }
+                }
+
+                return builder.ToImmutable();
+            }
+
+            // We didn't have regular source references, but possibly:
+            // 1. Another language (like XAML) will take over via ISymbolNavigationService
+            // 2. There are no locations from source, so we'll try to generate a metadata as source file and use that
+            var symbol = await SymbolFinder.FindSymbolAtPositionAsync(
+                document,
+                position,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            if (symbol == null)
+            {
+                return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+            }
+
+            var symbolNavigationService = workspace.Services.GetRequiredService<ISymbolNavigationService>();
+            var definitionItem = symbol.ToNonClassifiedDefinitionItem(document.Project.Solution, includeHiddenLocations: false);
+            var result = await symbolNavigationService.GetExternalNavigationSymbolLocationAsync(definitionItem, cancellationToken).ConfigureAwait(false);
+
+            if (result != null)
+            {
+                return ImmutableArray.Create(new CodeDefinitionWindowLocation(symbol.ToDisplayString(), result.Value.filePath, result.Value.linePosition));
+            }
+            else if (_metadataAsSourceFileService.IsNavigableMetadataSymbol(symbol))
+            {
+                // Don't allow decompilation when generating, since we don't have a good way to prompt the user
+                // without a modal dialog.
```

Comment:
Oh I still need to fix this.

Reply:
I didn't think we prompted any more

Reply:
We did when I wrote this comment months ago!

Reply:
@sharwell what's the current policy here? I see #52318 updated the VS symbol navigation service to pass allowDecompilation = true, but the rest of the call sites (LSP, peek, and now this) are still passing false. Was that an oversight or for perf concerns?

Reply:
Updated this spot; there's some more general cleanup here but have to sync with the VS for Mac team.